configs/provider_validation.go

```diff
@@ -0,0 +1,246 @@
+// module call, or a parent module call, has used a feature that precludes
+// providers from being configured at all within the module.
+func validateProviderConfigs(call *ModuleCall, cfg *Config, noProviderConfig bool) (diags hcl.Diagnostics) {
+	for name, child := range cfg.Children {
+		mc := cfg.Module.ModuleCalls[name]
+
+		// if the module call has any of count, for_each or depends_on,
+		// providers are prohibited from being configured in this module, or
+		// any module beneath this module.
+		nope := noProviderConfig || mc.Count != nil || mc.ForEach != nil || mc.DependsOn != nil
+		diags = append(diags, validateProviderConfigs(mc, child, nope)...)
+	}
+
+	// nothing else to do in the root module
+	if call == nil {
+		return diags
+	}
+
+	// the set of provider configuration names passed into the module, with the
+	// source range of the provider assignment in the module call.
+	passedIn := map[string]PassedProviderConfig{}
+
+	// the set of empty configurations that could be proxy configurations, with
+	// the source range of the empty configuration block.
+	emptyConfigs := map[string]*hcl.Range{}
+
+	// the set of provider with a defined configuration, with the source range
+	// of the configuration block declaration.
+	configured := map[string]*hcl.Range{}
+
+	// the set of configuration_aliases defined in the required_providers
+	// block, with the fully qualified provider type.
+	configAliases := map[string]addrs.AbsProviderConfig{}
+
+	// the set of provider names defined in the required_providers block, and
+	// their provider types.
+	localNames := map[string]addrs.AbsProviderConfig{}
+
+	for _, passed := range call.Providers {
+		name := providerName(passed.InChild.Name, passed.InChild.Alias)
+		passedIn[name] = passed
+	}
+
+	mod := cfg.Module
+
+	for _, pc := range mod.ProviderConfigs {
+		name := providerName(pc.Name, pc.Alias)
+		// Validate the config against an empty schema to see if it's empty.
+		_, pcConfigDiags := pc.Config.Content(&hcl.BodySchema{})
+		if pcConfigDiags.HasErrors() || pc.Version.Required != nil {
+			configured[name] = &pc.DeclRange
+		} else {
+			emptyConfigs[name] = &pc.DeclRange
+		}
+	}
+
+	if mod.ProviderRequirements != nil {
+		for _, req := range mod.ProviderRequirements.RequiredProviders {
+			addr := addrs.AbsProviderConfig{
+				Module:   cfg.Path,
+				Provider: req.Type,
+			}
+			localNames[req.Name] = addr
+			for _, alias := range req.Aliases {
+				addr := addrs.AbsProviderConfig{
+					Module:   cfg.Path,
+					Provider: req.Type,
+					Alias:    alias.Alias,
+				}
+				configAliases[providerName(alias.LocalName, alias.Alias)] = addr
+			}
+		}
+	}
+
+	// there cannot be any configurations if no provider config is allowed
+	if len(configured) > 0 && noProviderConfig {
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  fmt.Sprintf("Module %s contains provider configuration.", cfg.Path),
```

Comment:
```suggestion
			Summary:  fmt.Sprintf("Module %s contains provider configuration", cfg.Path),
```

I think we usually leave off that trailing `.` in summaries: 


```
Module module.child.module.child2 contains provider configuration.; Providers cannot be configured within modules using count, for_each or depends_on
```

Reply:
Thanks, I forgot how hcl diagnostics are slightly different anyway, and I'll follow up with better formatting in these.